### ISSUE TITLE
Invisible products do not update url_rewrite properly

### ISSUE THREAD

**Issue by dverkade** [magento/magento2#19293]
### Preconditions
1. Magento 2.4-develop

### Steps to reproduce
1. Have a product in the Magento backoffice with the visibility "Not visible individually". Set the URL key of the product to "product-a"
2. Open the product in de backoffice and change the URL Key to "product-b" and check the "create permanent redirect" checkbox
3. Save the product. 
4. Open the product again, change the visibility of the product to "Catalog, search" and change the URL key to "product-c", check the "create permanent redirect" checkbox.
5. Save the product again.

### Expected result
1. Permanent redirects should be present in the url rewrite database table from product-a to product-b
2. Permanent redirects should be present in the url rewrite database table from product-b to product-c
3. Expected result is that URL rewrite of products with "not individually visible" will also update.

### Actual result (*)
1. In the URL rewrite database table, only a references to product-c are available and permanent redirects from product-b to product-c are available. 
2. product-a is not redirecting to product-b
3. URL rewrites for products which where once visible and now are not visible anymore are not being updated anymore and the old URL key which was present when the product was still visible is being kept in the URL rewrite table.

**Comment by magento-engcom-team** [comment]
Hi @dverkade. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@dverkade do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by dverkade** [comment]
There are 2 possible solutions for this: 
- Either, when you set a product from visible to not visible the URL rewrites for that given product which are already in the URL rewrite table will be removed.
- We update the URL rewrite for the product whether the product is visible or not. 

Please advice about the best possible solution for this use case.

**Comment by orlangur** [comment]
> URL rewrites for products which where once visible and now are not visible anymore are not being updated anymore and the old URL key which was present when the product was still visible is being kept in the URL rewrite table.

@dverkade this is correct behavior, please adjust description accordingly. Redirects should work the way you described I believe.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @dverkade i'm not able to reproduce following steps you described, this is correct behavior, as mentioned a @orlangur , this is not a bug, if you like to update this please reopen the issue.

**Comment by dverkade** [comment]
I have created a screen capture to show you the issues and that the proces is not consequent. Hope this helps. 

Screen capture: https://www.flickr.com/photos/danny-anouk/44173445060

1. Create SKU "test product a" which is visible, URL key is saved in "url_rewrite" table (issue if item 4 is not an issue)
2. Create SKU "test product b" which is not visible, URL key is not saved in "url_rewrite" table
3. Switching visibility from SKU "test product b" to visible, URL key is saved in the "url_rewrite" table.
4. Switching visibility from SKU "test product b" to not visible, URL key is still present in "url_rewrite" table and is not removed (issue if item 1 is not an issue)
5. Changing URL key for SKU "test product a" to "test-product-b", product can't be saved because of URL key still present in the "url_rewrite" table.
6. Changing URL key for SKU "test product b" to "test-product-c". URL key is not updated in the "url_rewrite" table because the product is not visible. (issue, because when again doing item number 5 it still throws an error because the URL key is already taken)
7. Changing visibility from SKU "test product b" to visible. URL key is updated in the "url_rewrite" table and created a permanent redirect from the previous URL key.
8. Changing visibility from SKU "test product b" to not visible and changing the URL key to "test-product-d". I do not want to create a permanent redirect. However when we switch the product from not visible to visible again it creates a permanent redirect (issue, because I didn't want a permanent URL rewrite to be created, I just wanted to change the URL key)

**Comment by orlangur** [comment]
@engcom-backlog-nazar, I was only saying that

> URL rewrites for products which where once visible and now are not visible anymore are not being updated anymore and the old URL key which was present when the product was still visible is being kept in the URL rewrite table.

does not seem to be correct to me while two other points are valid.

**Comment by ghost** [comment]
> Changing visibility from SKU "test product b" to not visible and changing the URL key to "test-product-d". I do not want to create a permanent redirect. However when we switch the product from not visible to visible again it creates a permanent redirect (issue, because I didn't want a permanent URL rewrite to be created, I just wanted to change the URL key)

@orlangur  Yes now clear, this is not correct.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96573`, `MAGETWO-96574` were created

**Comment by dverkade** [comment]
@engcom-backlog-nazar @orlangur So this issue has been split into 2 different internal JIRA tickets. Can you give some pointers what is should be the intended behaviour of Magento, so that I can work on a PR to resolve this?

**Comment by orlangur** [comment]
@dverkade two issues are just for 2.2.x/2.3.x.

When product becomes invisible, URL Rewrite should remain in the database.

> URL key is not updated in the "url_rewrite" table because the product is not visible

Should be updated in this case, I believe.

**Comment by magento-engcom-team** [comment]
Hi @hardik-krish. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by dverkade** [comment]
> @dverkade two issues are just for 2.2.x/2.3.x.
> 
> When product becomes invisible, URL Rewrite should remain in the database.
> 
> > URL key is not updated in the "url_rewrite" table because the product is not visible
> 
> Should be updated in this case, I believe.

No, this is incorrect. When the product is created with visibility invisible the URL key is not created. It should be deleted when a product visibility is changed. Magento already approved / reviewed this requests, so no need to change the behaviour of this PR.

**Comment by engcom-Hotel** [comment]
Hello @dverkade.
I have checked the issue:

_Steps to reproduce from the description:
Have a product in the Magento back office with the visibility "Not visible individually". Set the URL key of the product to "product-a"
Open the product in de back-office and change the URL Key to "product-b" and check the "create permanent redirect" checkbox
Save the product.
Open the product again, change the visibility of the product to "Catalog, search" and change the URL key to "product-c", check the "create permanent redirect" checkbox.
Save the product again._

_Expected result
Permanent redirects should be present in the URL rewrite database table from product-a to product-b_
Why do we need redirects from invisible products? For what purposes it can be used? It is invisible means you are not able to open it via URL right?

_Permanent redirects should be present in the URL rewrite database table from product-b to product-c_
There is no point having 301 redirects from product-a URL to product-b since product-a was never accessible externally.

_The expected result is that URL rewrite of products with "not individually visible" will also update._
We need URL for invisible products only if it was visible before, otherwise, it shouldn’t have URL in url_rewrites table.

_Actual result (*)
In the URL rewrite database table, only references to product-c are available and permanent redirects from product-b to product-c are available._
Not accurate - URL rewrites functionality is available only for visible products, so, ‘_permanent redirects from product-b to product-c are available_’ part is not correct.

_product-a is not redirecting to product-b_
Correct, because it was invisible at that time.

_URL rewrites for products which were once visible and now are not visible anymore are not being updated anymore and the old URL key which was present when the product was still visible is being kept in the URL rewrite table._
That’s right, URL of the product which was visible and became invisible will be used for this product when it will be visible again/redirect from it to new url_key.
 
_Steps to reproduce from the comment (https://github.com/magento/magento2/issues/19293#issuecomment-440781945)
Create SKU "test product a" which is visible, URL key is saved in "url_rewrite" table (issue if item 4 is not an issue)_
Correct

_Create SKU "test product b" which is not visible, URL key is not saved in "url_rewrite" table_
Correct

_Switching visibility from SKU "test product b" to visible, URL key is saved in the "url_rewrite" table._
Correct

_Switching visibility from SKU "test product b" to not visible, URL key is still present in "url_rewrite" table and is not removed (issue if item 1 is not an issue)_
Correct (it could be used when it will be visible again)

_Changing the URL key for SKU "test product a" to "test-product-b", product can't be saved because of URL key still present in the "url_rewrite" table._
Correct

_Changing URL key for SKU "test product b" to "test-product-c". URL key is not updated in the "url_rewrite" table because the product is not visible. (issue, because when again doing item number 5 it still throws an error because the URL key is already taken)_
Correct, the product is invisible and can’t have an URL.

_Changing visibility from SKU "test product b" to visible. URL key is updated in the "url_rewrite" table and created a permanent redirect from the previous URL key._
Correct

_Changing visibility from SKU "test product b" to not visible and changing the URL key to "test-product-d". I do not want to create a permanent redirect. However, when we switch the product from not visible to visible again it creates a permanent redirect (issue, because I didn't want a permanent URL rewrite to be created, I just wanted to change the URL key)_
Not able to reproduce. If we change the url_key, remove ‘Create Permanent Redirect for old URL’ tick, make the product visible again and save - old record with the previous url_key will be removed.
 
Based on the provided analyzes I would recommend closing the issue because it works as expected.
 
The only one small bug with validation is that if you updated url_key for invisible product and save it - you are not able to remove ‘Create Permanent Redirect for old URL ’ tick until you don’t change the url_key. It is weird but not critical at all. The separate issue should be created about it if necessary.

Could you please review my comments and let me know if you have something to add.
Thank you.

**Comment by dverkade** [comment]
@engcom-Hotel, no need to test this as the issue has already been verified and reproduced.

**Comment by panteliskopsias** [comment]
Did you find any fix on this? 
I need the url rewrites in invisible products too. I need it because i export this urls in products feeds, ads and price comparison sites. 
From there when a user click on the product, he redirect to the configurable product with the option of simple preselected. 
The only way i can do it now is to create the products one by one with visibility 1,2 or 3 and then set it as not visible individually.
Thank you

**Comment by kanevbg** [comment]
up

**Comment by dverkade** [comment]
Found the following in the release notes for Magento 2.4.6: https://experienceleague.adobe.com/docs/commerce-operations/release/notes/magento-open-source/2-4-6.html?lang=en#url-rewrites 

> Changing product visibility to Not Visible Individually in the global All Store Views scope now removes only URL rewrites for store views that have not overridden the visibility attribute. Changing product visibility in this way also generates URL rewrites for all store views including those that have overridden the URL key attribute (but not the visibility setting). Previously, URL-rewrite rules were not applied when visibility settings changed.

Seems that this issue has been fixed @engcom-Hotel are you able to check the issue in the internal Jira instance and close this issue if it's been solved in Magento 2.4.6 and tag this accordingly?

**Comment by engcom-Hotel** [comment]
Hello @dverkade,

Reverified the issue in fresh Magento 2.4-develop installation. but it seems the issue is not resolved yet. In fresh magento installation w/o sample data, in the `url_rewrite` table, we have only 5 entries as below:

<img width="948" alt="image" src="https://user-images.githubusercontent.com/51681618/225659720-003d32bd-ac75-4b5b-b46e-7f31f62397e3.png">

Still, we are not able to see the expected result as mentioned in the main description.

Let us know if we have missed anything.

Thanks

**Comment by engcom-Bravo** [comment]
Hi @dverkade,

Thanks for your reporting and collaboration.

We have reverified the issue in Latest 2.4-develop instance and the issue is still reproducible.Kindly refer the screenshots.

<img width="878" alt="Screenshot 2024-10-23 at 11 02 25" src="https://github.com/user-attachments/assets/10ded20a-d0da-4fcd-82b0-e8042e950a28">


In the URL rewrite database table, only a references to product-c are available and permanent redirects from product-b to product-c are available.

Hence Confirming the issue.

Thanks.
